### PR TITLE
Update http(s).get and http(s).request libdefs

### PR DESCRIPTION
**Pull request by mrtnzlml**
Closes: https://github.com/facebook/flow/issues/7603

See:

- https://nodejs.org/api/http.html#http_http_get_url_options_callback
- https://nodejs.org/api/http.html#http_http_request_url_options_callback
- https://nodejs.org/api/https.html#https_https_get_url_options_callback
- https://nodejs.org/api/https.html#https_https_request_url_options_callback

### CODE REVIEW

**Comment by mvitousek** [comment]
Sorry it took a bit to get to this, it looks good. Is it possible though to get a more precise type for `options` than just `Object` (which is unsafe and generally should be avoided)?

**Comment by mrtnzlml** [comment]
@mvitousek Sure, will do. I was just extending what was already there a little bit. I'll let you know when it's ready to be checked again. :)

**Comment by mrtnzlml** [comment]
@mvitousek Updated and rebased. There are two things I am unsure about:

1) `http$requestOptions` and `https$requestOptions` are very similar with some small changes for HTTS - not sure if this is the best way how to do it. I would appreciate some guidance here if not.
2) I noticed that `createConnection` methods  in http(s) `Agent` are not compatible. I didn't touch this part at all in this PR but should probably be fixed. It's kind of related because there is also `createConnection` property in the request options I created.

Please, have a look and let me know what to improve.

File: lib/node.js [hunk 1]
```diff
@@ -1372,6 +1372,25 @@ declare class https$Server extends tls$Server {
   timeout: number;
 }
 
+type http$requestOptions = {
+  agent?: boolean | http$Agent<net$Socket>,
+  auth?: string,
+  createConnection?: (options: net$connectOptions, callback?: Function) => net$Socket,
+  defaultPort?: number,
+  family?: number,
+  headers?: {[key: string] : string},
```
Comment (mvitousek) [inline lib/node.js]: Unless I'm missing it the docs don't seem to specify that `headers` only maps strings to strings--seems like you could have at least numbers too. Maybe `{[key: string] : mixed}`?

**Comment by mvitousek** [comment]
Looks really good, thank you for taking this on! Made one suggestion, and since the request object is different only in its default values (which aren't visible by the caller), could you just use the same object for both?

Would you mind filing an issue regarding the `createConnection` definition?

**Comment by mrtnzlml** [comment]
> ... since the request object is different only in its default values (which aren't visible by the caller), could you just use the same object for both?

I tried to follow the differences between  HTTP and HTTPS module so there are additional differences in `http$Agent<net$Socket>` vs. `http$Agent<tls$TLSSocket>` and similar for return type of `createConnection`. Should it really be the same?

File: lib/node.js (hunk 1 above)
Comment (mrtnzlml) [reply lib/node.js]: ✅ Mixed it is.

**Comment by mvitousek** [comment]
Oops great catch, thank you. What I would do then--when most attributes are the same, but some are different--is to put the attributes in common in a single type, and then use spreading to copy those attributes into distinct types that also capture the differences. See here for an example: 
https://flow.org/try/#0C4TwDgpgBMULxQN4B8oEMBcUB2BXAtgEYQBOUyAvlAPTVQDKYJEaAJgJbYDmUA7gPYkA1gGcoxEbF7tgACygQAHmgDGsfoQBWENTHAQRAKFCR08JADorwADTiskkpx4Vj+8ecRWLt+zgLEZK6GrDoANmjMUABukVCKWGgA3IaxZNhYeESk5ooWyalxIg7ATty5FoRJQA

That said, if you'd prefer to not go down this route I'm happy to merge this as-is, as long as it passes all our tests.

**Comment by mrtnzlml** [comment]
@mvitousek Updated. Please have a look.

**Comment by mrtnzlml** [comment]
@mvitousek Hi, is there anything I can do to make it mergeable? :)

**Comment by mvitousek** [comment]
Hi @mrtnzlml, sorry for taking so long to finish this! The PR is landing right now :) Thanks a lot for your work on this PR, it's a great contribution and we appreciate it.

**Pull request closed without merging**